fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/DecorrelateValuesRule.java

```diff
@@ -0,0 +1,375 @@
+ * as:
+ * </p>
+ *
+ * <pre>{@code
+ * SELECT T.d
+ *   FROM T
+ *   WHERE a = 42 AND b = 'hello' AND c IS NULL
+ * }</pre>
+ */
+public class DecorrelateValuesRule extends ExplorationCascadesRule<SelectExpression> {
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> forEach = QuantifierMatchers.forEachQuantifier();
+
+    @Nonnull
+    private static final BindingMatcher<SelectExpression> root = RelationalExpressionMatchers.selectExpression(some(forEach));
+
+    public DecorrelateValuesRule() {
+        super(root);
+    }
+
+    private static <T> boolean emptyIntersection(@Nonnull Set<? extends T> set1, @Nonnull Set<? super T> set2) {
+        return set1.stream().noneMatch(set2::contains);
+    }
+
+    private static boolean correlatedToNone(@Nonnull RelationalExpression expression, @Nonnull Set<CorrelationIdentifier> ids) {
+        return emptyIntersection(expression.getCorrelatedTo(), ids);
+    }
+
+    private static boolean correlatedToNone(@Nonnull Quantifier qun, @Nonnull Set<CorrelationIdentifier> ids) {
+        return emptyIntersection(qun.getCorrelatedTo(), ids);
+    }
+
+    @Override
+    public void onMatch(@Nonnull final ExplorationCascadesRuleCall call) {
+        final SelectExpression selectExpression = call.get(root);
+        final List<? extends Quantifier.ForEach> valueQunCandidates = call.getBindings().getAll(forEach);
+
+        final Set<CorrelationIdentifier> selectChildQunIds = selectExpression.getQuantifiers().stream().map(Quantifier::getAlias).collect(ImmutableSet.toImmutableSet());
+        ImmutableMap.Builder<CorrelationIdentifier, SelectExpression> valuesByIdBuilder = ImmutableMap.builderWithExpectedSize(valueQunCandidates.size());
+        ImmutableMap.Builder<CorrelationIdentifier, Quantifier> qunsByIdBuilder = ImmutableMap.builderWithExpectedSize(valueQunCandidates.size());
+        for (Quantifier.ForEach qun : valueQunCandidates) {
+            @Nullable SelectExpression childSelect = findSelectForQuantifier(qun, selectChildQunIds);
+            if (childSelect != null) {
+                valuesByIdBuilder.put(qun.getAlias(), childSelect);
+                qunsByIdBuilder.put(qun.getAlias(), qun);
+            }
+        }
+        final Map<CorrelationIdentifier, SelectExpression> valuesById = valuesByIdBuilder.build();
+        if (valuesById.isEmpty()) {
+            // No actual values boxes here. Exit now
+            return;
+        }
+        if (valuesById.size() == selectExpression.getQuantifiers().size()) {
+            // All the quantifiers are values boxes. We can't push them around as this would
+            // leave no children at all. So we exit here.
+            return;
+        }
+
+        //
+        // Create a translation map, and use it to translate the result value of this select as well as any predicates
+        //
+        final TranslationMap translationMap = createTranslationMapFromSelects(valuesById);
```

Comment:
I think what we discussed earlier makes sense to me now, i.e. inline the values for the current expression and not immediately for the select your are potentially pushing into. That can be job of the next invocation. I think that's fine.